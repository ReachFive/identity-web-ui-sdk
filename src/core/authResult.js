import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
