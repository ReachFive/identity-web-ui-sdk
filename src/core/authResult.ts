import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
