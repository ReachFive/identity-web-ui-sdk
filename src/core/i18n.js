import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
