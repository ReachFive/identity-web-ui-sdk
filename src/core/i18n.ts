export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
