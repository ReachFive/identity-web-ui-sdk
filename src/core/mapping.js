import { getValue, setValue } from '../helpers/propertyHelpers';

export class PathMapping {
    constructor(modelPath) {
        this.modelPath = modelPath;
    }

    bind(model) {
        return getValue(model, this.modelPath);
    }

    unbind(model, value) {
        return setValue(model, this.modelPath, value);
    }
}
