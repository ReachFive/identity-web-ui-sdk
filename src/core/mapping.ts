import { ProfileAddress } from '@reachfive/identity-core';
import { getValue, setValue } from '../helpers/propertyHelpers';

export interface PathMapping {
    bind<T extends Record<string, unknown>>(model: T): unknown;
    unbind<T extends Record<string, unknown>, V>(model: T, value: V): T | V;
}

export class DefaultPathMapping implements PathMapping {
    protected readonly modelPath: string;

    constructor(modelPath: string) {
        this.modelPath = modelPath;
    }

    bind<T extends Record<string, unknown>>(model: T) {
        return getValue(model, this.modelPath);
    }

    unbind<T extends Record<string, unknown>, V>(model: T, value: V) {
        return setValue(model, this.modelPath, value);
    }
}

export class AddressPathMapping implements PathMapping {
    protected readonly property: string;

    constructor(property: string) {
        this.property = property;
    }

    private defaultIndex(model: { addresses?: ProfileAddress[] }) {
        const defaultIndex = (model.addresses ?? []).findIndex(address => address.isDefault);
        return defaultIndex > -1 ? defaultIndex : 0;
    }

    bind<T extends { addresses?: ProfileAddress[] }>(model: T) {
        return getValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`);
    }

    unbind<T extends { addresses?: ProfileAddress[] }, V>(model: T, value: V) {
        return setValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`, value);
    }
}
