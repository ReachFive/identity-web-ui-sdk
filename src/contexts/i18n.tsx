import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
