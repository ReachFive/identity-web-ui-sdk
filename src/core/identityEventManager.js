import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
