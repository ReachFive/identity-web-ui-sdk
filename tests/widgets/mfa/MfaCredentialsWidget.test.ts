/**
 * @jest-environment jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Client, MFA } from '@reachfive/identity-core'

import { I18nMessages } from '../../../src/core/i18n';
import type { Config } from '../../../src/types';

import mfaCredentialsWidget from '../../../src/widgets/mfa/MfaCredentialsWidget';

const defaultConfig: Config = {
    clientId: 'local',
    domain: 'local.reach5.net',
    sso: false,
    sms: false,
    webAuthn: false,
    language: 'fr',
    pkceEnforced: false,
    isPublic: true,
    socialProviders: ['facebook', 'google'],
    customFields: [],
    resourceBaseUrl: 'http://localhost',
    mfaSmsEnabled: true,
    mfaEmailEnabled: true,
    rbaEnabled: false,
    consentsVersions: {},
    passwordPolicy: {
        minLength: 8,
        minStrength: 2,
        allowUpdateWithAccessTokenOnly: true,
    },
};

const defaultI18n: I18nMessages = {};
const profile = {
    emailVerified: true,
    authTypes: [],
    emails: {
        verified: [],
        unverified: [],
    },
    thirdPartyGrants: [],
    tokenRevocationRecord: {
        allLongLived: undefined,
        longLivedByClient: {},
    },
    providers: [],
    likesFriendsRatio: 0,
    localFriendsCount: 0,
    loginsCount: 0,
    origins: [],
    devices: [],
    hasPassword: false,
    socialIdentities: [],
    hasManagedProfile: false,
    providerDetails: [],
};

describe('Snapshot', () => {
    const generateSnapshot =
        (
            options: Partial<Parameters<typeof mfaCredentialsWidget>[0]>,
            config: Partial<Config> = {},
            credentials: (MFA.EmailCredential | MFA.PhoneCredential)[]
        ) =>
        async () => {
            // @ts-expect-error partial Client
            const apiClient: Client = {
                listMfaCredentials: jest
                    .fn<Client['listMfaCredentials']>()
                    .mockResolvedValue({ credentials }),
                getUser: jest.fn<Client['getUser']>().mockResolvedValue(profile),
            };

            const widget = await mfaCredentialsWidget(
                { accessToken: 'azerty', showIntro: true, ...options },
                { apiClient, config: { ...defaultConfig, ...config }, defaultI18n }
            );

            await waitFor(async () => {
                const { container, rerender } = await render(widget);

                await waitFor(() => expect(apiClient.listMfaCredentials).toHaveBeenCalled());

                await rerender(widget);

                expect(container).toMatchSnapshot();
            });
        };

    describe('mfaCredentials', () => {
        test('default', generateSnapshot({}, undefined, []))

        test(
            'no intro',
            generateSnapshot({ showIntro: false }, undefined, [
                {
                    type: 'sms',
                    phoneNumber: '33612345678',
                    friendlyName: 'identifier',
                    createdAt: '2022-09-21',
                },
                {
                    type: 'email',
                    email: 'root@reach5.co',
                    friendlyName: 'identifier',
                    createdAt: '2022-09-21',
                },
            ])
        );
    });
});

describe('DOM testing', () => {
    const listMfaCredentials = jest.fn<Client['listMfaCredentials']>();
    const startMfaEmailRegistration = jest.fn<Client['startMfaEmailRegistration']>();
    const startMfaPhoneNumberRegistration = jest.fn<Client['startMfaPhoneNumberRegistration']>();
    const verifyMfaEmailRegistration = jest.fn<Client['verifyMfaEmailRegistration']>();
    const verifyMfaPhoneNumberRegistration = jest.fn<Client['verifyMfaPhoneNumberRegistration']>();
    const getUser = jest.fn<Client['getUser']>();

    const onError = jest.fn();
    const onSuccess = jest.fn();

    beforeEach(() => {
        onError.mockClear();
        onSuccess.mockClear();
        listMfaCredentials.mockClear();
        startMfaEmailRegistration.mockClear();
        startMfaPhoneNumberRegistration.mockClear();
        verifyMfaEmailRegistration.mockClear();
        verifyMfaPhoneNumberRegistration.mockClear();
        getUser.mockClear();
    });

    const generateComponent = async (
        options: Partial<Parameters<typeof mfaCredentialsWidget>[0]>,
        config: Partial<Config> = {},
        credentials: (MFA.EmailCredential | MFA.PhoneCredential)[]
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            listMfaCredentials: listMfaCredentials.mockResolvedValue({ credentials }),
            startMfaEmailRegistration: startMfaEmailRegistration.mockResolvedValue({
                status: 'email_sent',
            }),
            startMfaPhoneNumberRegistration: startMfaPhoneNumberRegistration.mockResolvedValue({
                status: 'sms_sent',
            }),
            verifyMfaEmailRegistration: verifyMfaEmailRegistration.mockResolvedValue(),
            verifyMfaPhoneNumberRegistration: verifyMfaEmailRegistration.mockResolvedValue(),
            getUser: getUser.mockResolvedValue(profile),
        };
        const result = await mfaCredentialsWidget(
            {
                accessToken: 'azerty',
                onError,
                onSuccess,
                ...options,
            },
            {
                apiClient,
                config: { ...defaultConfig, ...config },
                defaultI18n,
            }
        );
        return await waitFor(async () => render(result));
    };

    describe('mfaCredentials', () => {
        test('no credentials', async () => {
            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true },
                defaultConfig,
                []
            );
            // Email intro
            expect(screen.queryByText('mfa.email.explain')).toBeInTheDocument()

            // Form button email
            expect(screen.queryByText('mfa.register.email')).toBeInTheDocument()

            // Sms intro
            expect(
                screen.queryByText('mfa.phoneNumber.explain')
            ).toBeInTheDocument()

            // Form button sms
            expect(
                screen.queryByText('mfa.register.phoneNumber')
            ).toBeInTheDocument()

            // Form button remove email
            expect(
                screen.queryByText('mfa.remove.email')
            ).not.toBeInTheDocument()

            // Form button remove phone number
            expect(
                screen.queryByText('mfa.remove.phoneNumber')
            ).not.toBeInTheDocument()
        })

        test('register email', async () => {
            const user = userEvent.setup();

            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true },
                defaultConfig,
                []
            );

            // Form button email
            const emailButton = screen.getByText('mfa.register.email');
            expect(emailButton).toBeInTheDocument();

            await user.click(emailButton);
            expect(startMfaEmailRegistration).toBeCalled();

            await waitFor(async () => {
                expect(screen.queryByText('mfa.verify.email')).toBeInTheDocument();
            });

            const verificationCodeInput = screen.getByTestId('verification_code');
            expect(verificationCodeInput).toBeInTheDocument();
            await user.type(verificationCodeInput, '123456');

            verifyMfaEmailRegistration.mockReset().mockRejectedValue(new Error('Invalid code'));

            await user.click(screen.getByTestId('submit'));

            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalled();

            verifyMfaEmailRegistration.mockReset().mockResolvedValue();
            onSuccess.mockReset();
            onError.mockReset();

            await user.click(screen.getByTestId('submit'));

            await waitFor(async () => {
                expect(screen.queryByText('mfa.email.registered')).toBeInTheDocument();
                expect(onSuccess).toBeCalled();
                expect(onError).not.toBeCalled();
            });
        });

        test('register phone number', async () => {
            const user = userEvent.setup();

            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true },
                defaultConfig,
                []
            );

            // phone number input
            const phoneNumberInput = screen.getByTestId('phone_number');
            await user.type(phoneNumberInput, '+33123456789');

            // Form button phone number
            const phoneNumberButton = screen.getByText('mfa.register.phoneNumber');
            expect(phoneNumberButton).toBeInTheDocument();

            await user.click(phoneNumberButton);
            expect(startMfaPhoneNumberRegistration).toBeCalled();

            await waitFor(async () => {
                expect(screen.queryByText('mfa.verify.sms')).toBeInTheDocument();
            });

            const verificationCodeInput = screen.getByTestId('verification_code');
            expect(verificationCodeInput).toBeInTheDocument();
            await user.type(verificationCodeInput, '123456');
            await user.click(screen.getByTestId('submit'));

            await waitFor(async () => {
                expect(screen.queryByText('mfa.phoneNumber.registered')).toBeInTheDocument();
                expect(onSuccess).toBeCalled();
                expect(onError).not.toBeCalled();
            });
        });

        test('requireMfaRegistration', async () => {
            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true, requireMfaRegistration: true },
                defaultConfig,
                []
            );
            // Email intro
            expect(
                screen.queryByText('mfa.email.explain.required')
            ).toBeInTheDocument()

            // Form button email
            expect(screen.queryByText('mfa.register.email')).toBeInTheDocument()

            // Sms intro
            expect(
                screen.queryByText('mfa.phoneNumber.explain')
            ).toBeInTheDocument()

            // Form button sms
            expect(
                screen.queryByText('mfa.register.phoneNumber')
            ).toBeInTheDocument()

            // Form button remove email
            expect(
                screen.queryByText('mfa.remove.email')
            ).not.toBeInTheDocument()

            // Form button remove phone number
            expect(
                screen.queryByText('mfa.remove.phoneNumber')
            ).not.toBeInTheDocument()
        })

        test('only email credential', async () => {
            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true },
                defaultConfig,
                [
                    {
                        type: 'email',
                        email: 'alice@reach5.co',
                        friendlyName: 'identifier',
                        createdAt: '2022-09-21',
                    },
                ]
            );
            // Email intro
            expect(
                screen.queryByText('mfa.email.explain')
            ).not.toBeInTheDocument()

            // Form button email
            expect(
                screen.queryByText('mfa.register.email')
            ).not.toBeInTheDocument()

            // Sms intro
            expect(
                screen.queryByText('mfa.phoneNumber.explain')
            ).toBeInTheDocument()

            // Form button sms
            expect(
                screen.queryByText('mfa.register.phoneNumber')
            ).toBeInTheDocument()

            // Form button remove email
            expect(screen.queryByText('mfa.remove.email')).toBeInTheDocument()

            // Form button remove phone number
            expect(
                screen.queryByText('mfa.remove.phoneNumber')
            ).not.toBeInTheDocument()
        })

        test('only sms credential', async () => {
            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true },
                defaultConfig,
                [
                    {
                        type: 'sms',
                        phoneNumber: '33612345678',
                        friendlyName: 'identifier',
                        createdAt: '2022-09-21',
                    },
                ]
            );
            // Email intro
            expect(screen.queryByText('mfa.email.explain')).toBeInTheDocument()

            // Form button email
            expect(screen.queryByText('mfa.register.email')).toBeInTheDocument()

            // Sms intro
            expect(
                screen.queryByText('mfa.phoneNumber.explain')
            ).not.toBeInTheDocument()

            // Form button sms
            expect(
                screen.queryByText('mfa.register.phoneNumber')
            ).not.toBeInTheDocument()

            // Form button remove email
            expect(
                screen.queryByText('mfa.remove.email')
            ).not.toBeInTheDocument()

            // Form button remove phone number
            expect(
                screen.queryByText('mfa.remove.phoneNumber')
            ).toBeInTheDocument()
        })

        test('all credentials', async () => {
            await generateComponent(
                { showIntro: true, showRemoveMfaCredentials: true },
                defaultConfig,
                [
                    {
                        type: 'sms',
                        phoneNumber: '33612345678',
                        friendlyName: 'identifier',
                        createdAt: '2022-09-21',
                    },
                    {
                        type: 'email',
                        email: 'root@reach5.co',
                        friendlyName: 'identifier',
                        createdAt: '2022-09-21',
                    },
                ]
            );
            // Intro
            expect(
                screen.queryByText('mfa.email.explain')
            ).not.toBeInTheDocument()

            // Form button email
            expect(
                screen.queryByText('mfa.register.email')
            ).not.toBeInTheDocument()

            // Sms intro
            expect(
                screen.queryByText('mfa.phoneNumber.explain')
            ).not.toBeInTheDocument()

            // Form button sms
            expect(
                screen.queryByText('mfa.register.phoneNumber')
            ).not.toBeInTheDocument()

            // Form button remove email
            expect(screen.queryByText('mfa.remove.email')).toBeInTheDocument()

            // Form button remove phone number
            expect(
                screen.queryByText('mfa.remove.phoneNumber')
            ).toBeInTheDocument()
        })
    })
})
