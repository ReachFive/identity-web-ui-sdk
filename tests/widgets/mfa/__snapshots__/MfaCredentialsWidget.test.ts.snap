// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot mfaCredentials default 1`] = `
<div>
  <div
    class="sc-imWYAI fLZoZn sc-dhKdcB lapPam r5-widget"
    transition="entering"
  >
    <div
      class="flex flex-col gap-4"
    >
      <div
        id="email-registering-credential"
      >
        <div
          class="sc-gEvEer sc-eqUAAy bEpsus iRHKeq"
        >
          mfa.email.explain
        </div>
        <form
          class="sc-ikkxIA dcGBJp"
          novalidate=""
        >
          <button
            class="sc-kpDqfm izDDoy"
            data-testid="submit"
            type="submit"
          >
            mfa.register.email
          </button>
        </form>
      </div>
      <div
        class="sc-iGgWBj fIbmkT"
      >
        <span>
          or
        </span>
      </div>
      <div
        id="phone-number-registering-credential"
      >
        <div
          class="sc-gEvEer sc-eqUAAy bEpsus iRHKeq"
        >
          mfa.phoneNumber.explain
        </div>
        <form
          class="sc-ikkxIA dcGBJp"
          novalidate=""
        >
          <div
            class="sc-jEACwC liiTZm"
            required=""
          >
            <label
              class="sc-jlZhew jwHbOB"
              for="phone_number"
            >
              phoneNumber
            </label>
            <input
              autocomplete="tel"
              class="sc-cPiKLX bxqNC"
              data-testid="phone_number"
              id="phone_number"
              labels="[object Object]"
              name="phone_number"
              placeholder="phoneNumber"
              required=""
              type="tel"
              value=""
            />
          </div>
          <button
            class="sc-kpDqfm izDDoy"
            data-testid="submit"
            type="submit"
          >
            mfa.register.phoneNumber
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot mfaCredentials no intro 1`] = `
<div>
  <div
    class="sc-imWYAI fLZoZn sc-dhKdcB lapPam r5-widget"
    transition="entering"
  >
    <div
      class="flex flex-col gap-4"
    >
      <div>
        <form
          class="sc-ikkxIA dcGBJp"
          novalidate=""
        >
          <button
            class="sc-kpDqfm kpWHYP"
            type="submit"
          >
            mfa.remove.email
          </button>
        </form>
      </div>
      <div
        class="sc-iGgWBj fIbmkT"
      >
        <span>
          or
        </span>
      </div>
      <div>
        <form
          class="sc-ikkxIA dcGBJp"
          novalidate=""
        >
          <button
            class="sc-kpDqfm kpWHYP"
            type="submit"
          >
            mfa.remove.phoneNumber
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
